### PR TITLE
keyspace, api: support the keyspace group split

### PR DESCRIPTION
**Pull request by JmPotato**
### What problem does this PR solve?

Issue Number: ref #6232.

### What is changed and how does it work?

```commit-message
Support the keyspace group split and add related tests.
```

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/keyspace/tso_keyspace_group.go [hunk 1]
```diff
@@ -81,38 +81,146 @@ func (m *GroupManager) GetKeyspaceGroupByID(id uint32) (*endpoint.KeyspaceGroup,
...
+		// Get the split keyspace group for the old keyspace group.
+		splitKeyspaces := make([]uint32, 0, len(oldKg.Keyspaces)-len(keyspaces))
+		for _, keyspace := range oldKg.Keyspaces {
+			if _, ok := newKeyspaceMap[keyspace]; !ok {
+				splitKeyspaces = append(splitKeyspaces, keyspace)
+			}
+		}
+		// Update the old keyspace group.
+		oldKg.Keyspaces = splitKeyspaces
+		m.store.SaveKeyspaceGroup(txn, oldKg)
```
Comment (binshi-bing) [inline pkg/keyspace/tso_keyspace_group.go:196]: if len(oldKg.Keyspaces) == 0, shall we delete the old keyspace group?

File: server/apiv2/handlers/tso_keyspace_group.go [hunk 2]
```diff
@@ -122,6 +126,55 @@ func DeleteKeyspaceGroupByID(c *gin.Context) {
...
+		return
+	}
+	if !isValid(splitParams.ID) || !isValid(splitParams.NewID) {
+		c.AbortWithStatusJSON(http.StatusBadRequest, "invalid keyspace group id")
+		return
+	}
+
+	svr := c.MustGet(middlewares.ServerContextKey).(*server.Server)
+	manager := svr.GetKeyspaceGroupManager()
+	err = manager.SplitKeyspaceGroupByID(splitParams.ID, splitParams.NewID, splitParams.Keyspaces)
```
Comment (binshi-bing) [inline server/apiv2/handlers/tso_keyspace_group.go:152]: if len(splitParams.Keyspaces) == 0, we return earlier.

File: pkg/keyspace/tso_keyspace_group.go (hunk 1 above)
Comment (binshi-bing) [inline pkg/keyspace/tso_keyspace_group.go:196]: check error

File: pkg/keyspace/tso_keyspace_group.go [hunk 3]
```diff
@@ -81,38 +81,146 @@ func (m *GroupManager) GetKeyspaceGroupByID(id uint32) (*endpoint.KeyspaceGroup,
...
+		// Create the new split keyspace group.
+		return m.store.SaveKeyspaceGroup(txn, &endpoint.KeyspaceGroup{
```
Comment (binshi-bing) [inline pkg/keyspace/tso_keyspace_group.go:198]: Though, the put events to the old and new groups will be propagated asynchronously, we should still update/create both keyspace groups in a transaction to provide atomic operation.

File: pkg/keyspace/tso_keyspace_group.go [hunk 4]
```diff
@@ -81,38 +81,146 @@ func (m *GroupManager) GetKeyspaceGroupByID(id uint32) (*endpoint.KeyspaceGroup,
...
+			ID: newID,
+			// Keep the same user kind and members as the old keyspace group.
+			UserKind:  oldKg.UserKind,
+			Members:   oldKg.Members,
+			Keyspaces: keyspaces,
+			// Only set the new keyspace group in split state.
+			InSplit: true,
```
Comment (binshi-bing) [inline pkg/keyspace/tso_keyspace_group.go:205]: Do we need one more field to denote this keyspace group is split from which keyspace group?

File: server/apiv2/handlers/tso_keyspace_group.go [hunk 5]
```diff
@@ -122,6 +126,55 @@ func DeleteKeyspaceGroupByID(c *gin.Context) {
...
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, err.Error())
+		return
+	}
+	c.JSON(http.StatusOK, nil)
+}
+
+// FinishSplitKeyspaceByID finishes split keyspace group by ID.
+func FinishSplitKeyspaceByID(c *gin.Context) {
```
Comment (binshi-bing) [inline server/apiv2/handlers/tso_keyspace_group.go:161]: Who calls FinishSplitKeyspaceByID -- tso.KeyspaceGroupManager or human? For the latter, how do we know the split is successfully completed on the tso.KeyspaceGroupManager side?

**Review by binshi-bing — COMMENTED** [review]
left some comments

File: pkg/keyspace/tso_keyspace_group.go (hunk 3 above)
Comment (JmPotato) [reply pkg/keyspace/tso_keyspace_group.go:198]: The whole `SplitKeyspaceGroupByID` is in a txn.

File: pkg/keyspace/tso_keyspace_group.go (hunk 4 above)
Comment (JmPotato) [reply pkg/keyspace/tso_keyspace_group.go:205]: At present, I haven't thought of any scenarios where this field would be useful. Perhaps we can add it when we actually need it.

File: server/apiv2/handlers/tso_keyspace_group.go (hunk 5 above)
Comment (JmPotato) [reply server/apiv2/handlers/tso_keyspace_group.go:165]: It will be used by `tso.KeyspaceGroupManager` to finish the split.

**Review by binshi-bing — APPROVED** [review]
LGTM

File: server/apiv2/handlers/tso_keyspace_group.go [hunk 6]
```diff
@@ -35,6 +35,8 @@ func RegisterTSOKeyspaceGroup(r *gin.RouterGroup) {
 	router.GET("", GetKeyspaceGroups)
 	router.GET("/:id", GetKeyspaceGroupByID)
 	router.DELETE("/:id", DeleteKeyspaceGroupByID)
+	router.POST("/split", SplitKeyspaceGroupByID)
```
Comment (rleungx) [inline server/apiv2/handlers/tso_keyspace_group.go:38]: No need `:id`? And how about `/:id/split`

Comment (rleungx) [reply server/apiv2/handlers/tso_keyspace_group.go:38]: ref: https://docs.github.com/en/rest/gists/gists?apiVersion=2022-11-28#fork-a-gist

**Review by rleungx — APPROVED** [review]
The rest LGTM

**Pull request merged**